protocol/morse/sanctioned_endpoints_store.go

```diff
@@ -185,10 +187,101 @@ func (ses *sanctionedEndpointsStore) isSanctioned(
 	return false, ""
 }
 
-// TODO_MVP(@adshmh): return an error if any of the composite key components are empty.
-//
+// --------- Session Sanction Key ---------
+
 // sessionSanctionKey creates a unique key for session-based sanctions
 // Format: app_addr:session_key:endpoint_addr
-func sessionSanctionKey(appAddr, sessionKey, endpointAddr string) string {
-	return fmt.Sprintf("%s:%s:%s", appAddr, sessionKey, endpointAddr)
+type sessionSanctionKey struct {
+	appAddr      string
+	sessionKey   string
+	endpointAddr protocol.EndpointAddr
+}
+
+// TODO_MVP(@adshmh): return an error if any of the composite key components are empty.
+func newSessionSanctionKey(appAddr, sessionKey string, endpointAddr protocol.EndpointAddr) sessionSanctionKey {
+	return sessionSanctionKey{
+		appAddr:      appAddr,
+		sessionKey:   sessionKey,
+		endpointAddr: endpointAddr,
+	}
+}
+
+func newSessionSanctionKeyFromCacheKey(cacheKey string) sessionSanctionKey {
+	parts := strings.Split(cacheKey, ":")
+	return sessionSanctionKey{
+		appAddr:      parts[0],
+		sessionKey:   parts[1],
+		endpointAddr: protocol.EndpointAddr(parts[2]),
+	}
+}
+
+func (s sessionSanctionKey) string() string {
+	return fmt.Sprintf("%s:%s:%s", s.appAddr, s.sessionKey, s.endpointAddr)
+}
+
+// --------- Sanction Details ---------
+
+// getSanctionDetails returns the sanctioned endpoints for a given service ID.
+// It provides information about:
+//   - the currently sanctioned endpoints, including the reason
+//   - counts for valid and sanctioned endpoints
+//
+// It is called by the router to allow quick information about currently sanctioned endpoints.
+// This will eventually be removed in favour of a metrics-based approach.
+func (ses *sanctionedEndpointsStore) getSanctionDetails(serviceID protocol.ServiceID) devtools.SanctionDetailsResponse {
+	permanentSanctionDetails := make(map[protocol.ServiceID][]devtools.SanctionDetails)
+	permanentSanctionedEndpointsCount := 0
+
+	sessionSanctionDetails := make(map[protocol.ServiceID][]devtools.SanctionDetails)
+	sessionSanctionedEndpointsCount := 0
+
+	// First get permanent sanctions
+	for endpointAddr, sanction := range ses.permanentSanctions {
```

Comment:
Can you split this into smaller helpers with more detailed comments? It'll be easier to read as well.